### DISCUSSION TITLE
Named routes/Aliasing system for easier navigation

### DISCUSSION THREAD

**Discussion by ferretwithaberet** [TanStack/router#2456] [Ideas]
Other routers (e.g. [Vue Router](https://router.vuejs.org/guide/essentials/named-routes.html)) have the option of specifing a `name` property in the route options, this allows us to do stuff like `navigate({ name: "profile", params: { userId: 1 } })` instead of `navigate({ to: "/account/profile", params: { userId: 1 } })`.

The claimed benefits of this, from the above Vue Router link are:
- No hardcoded URLs.
- Automatic encoding of params.
- Avoids URL typos.
- Bypassing path ranking, e.g. to display a lower-ranked route that matches the same path.

The main benefit for me is no hardcoded URLs, as then we could move routes around without having to update `navigate()` calls.

**Comment by schiller-manuel** [comment]
Thanks for raising this idea!

I discussed this with @chorobin, here is a quick summary:

- as you pointed out already, the only real benefit is "no hardcoded urls", everything else is already in place.
- tanstack router already has a similar concept to `name`, you can give a route a custom `id`
- however, file based routes do not support this custom `id` as the `id` is auto generated.
- for this to work with file based routes, code generator would need to read the route definition AST and extract the `id`, and use it instead of auto generating it
- with virtual file based routes it could be easier to support this
- additionally, runtime would need to support this. most likely, another property than `to` would be needed to express you want to navigate based on an `id` instead of a path.
- relative paths would probably not be supported

This needs more thought put into. As of now, this would be "nice to have" but not a priority feature.
